0726-dom-element-descriptor-interface.md

```diff
@@ -0,0 +1,228 @@
+2. It does not produce useful information when encountering an error/failure. For example, `await click('.does-not-exist')` errors with `Error: Element not found when calling click('.does-not-exist').`, while `await click(document.querySelector('.does-not-exist'))` errors with the less helpful `Error: Must pass an element or selector to 'click'` (because it doesn't have any other information to show -- all it knows is it was passed `null` instead of a selector or `Element`). Similarly, `assert.dom('.does-not-exist').exists()` produces the assertion message `Element .does-not-exist exists`, while `assert.dom(document.querySelector('.does-not-exist')).exists()` produces the less helpful `Element <not found> exists`.
+3. It supports a noticeably less ergonomic integration with page objects. In `fractal-page-object`, the integrations currently look like `await click(page.listItems[2].checkbox.element)` and `assert.dom(page.listItems[2].checkbox.element).isChecked()`, rather than `await click(page.listItems[2].checkbox)` and `assert.dom(page.listItems[2].checkbox).isChecked()`. While it may look like a minor difference, in practice it's quite a stumbling block as the mental model of page objects encourages users to think of page object nodes (e.g. `page.listItems[2].checkbox`) as a proxy for the DOM element(s) they describe, so forgetting to add the `.element` is extremely common, even for the author of the library!
+
+We can solve these by formalizing the notion of DOM element descriptors in an interface, so page object implementations and any other mechanisms for DOM access can produce objects implementing this interface and they can be passed to these test helpers without incurring any of the shortcoming described above.
+
+## Detailed design
+
+### New use cases
+
+We're seeking to address two specific new use cases, although the design should be extensible to other unforeseen use cases as well.
+
+#### Page objects
+
+Page objects can be used as DOM element descriptors and passed directly to DOM helpers:
+
+```js
+assert.dom(pageObject.listItems).exists({ count: 4 });
+
+await click(pageObject.listItems[2].checkbox);
+
+assert.dom(pageObject.listItems[2].checkbox).isChecked();
+```
+
+This will enable more ergonomic integrations of libraries like `ember-cli-page-object` and `fractal-page-object` with DOM helpers.
+
+#### Ad-hoc descriptors
+
+DOM element descriptors can be constructed directly from `Element`s:
+
+```ts
+let element = someOtherLibrary.getGraphElement();
+let descriptor = createDOMDescriptor({ element, description: 'graph element' });
+
+await click(descriptor);
+
+assert.dom(descriptor).hasClass('selected');
+```
+
+or from simple classes for lazy DOM lookups:
+
+```ts
+class MyDescriptor {
+  get element() {
+    return document.querySelectorAll('.list-item')[2];
+  }
+  description = 'second list item';
```

Comment:
I guess we can't have these properties defined directly on the descriptor, cause like stated in the `Namespacing constraints` section, it may lead to names collisions with the user declarations? And seems like it's not compatible with the descriptors produced by the `createDOMDescriptor`?

Reply:
Whoops, you're totally right, good catch! I just messed up the syntax of that example -- it's fixed now I think.